### ISSUE TITLE
Locale setup is incomplete, LANG is not configured

### ISSUE THREAD

**Issue by tilldoerges** [SUSE/kiwi#1081]
## Problem description
The default locale for users created differs from building under openSUSE Leap 42.3 to building under openSUSE Leap 15.0.

The guest OS in both cases is 42.3.

This is the problematic case (build host Leap 15.0):
```
builton150:~ # su - user
user@builton150:~> locale
LANG=
LC_CTYPE="POSIX"
[...]
```

I'm guessing this issue is somewhat related to #985.

## Expected behaviour
This is how it previously was (build host Leap 42.3):
```
builton423~ # su - user
user@builton423~> locale
LANG=de_DE.UTF-8
LC_CTYPE="de_DE.UTF-8"
[...]
```

## Steps to reproduce the behaviour
Generate an ISO for Leap 42.3 under a Leap 15.0 build host.

## OS and Software information
+ KIWI version: 9.17.31
+ Operating system: Build host: openSUSE Leap 15.0, guest host: openSUSE Leap 42.3
+ OBS version: n.a.

**Comment by davidcassany** [comment]
@tilldoerges sorry I have several questions to fully narrow your current situation.

Have you done the built with the same KIWI version (9.17.31) in both hosts?

What you mean by the default locale? Is your config.xml not including any locale setting? Or you mean that the locale is not being set to what is stated in `<locale>`?

Apart from the different behavior is there anything that is specially causing you an issue? Have you tested the output of `localectl` command? 

Currently locale, keyboard and other related configurations are handled by systemd, so I would expect that the `localectl` returns you the correct value and that system behaves accordingly to systemd.

**Comment by tilldoerges** [comment]
>@tilldoerges sorry I have several questions to fully narrow your current situation.

Thanks for getting back so quickly.

> Have you done the built with the same KIWI version (9.17.31) in both hosts?

Build host on openSUSE Leap 15.0: Kiwi 9.17.31
Build host on openSUSE Leap 42.3: Kiwi 9.7.4

> What you mean by the default locale? Is your config.xml not including any locale setting? Or you mean that the locale is not being set to what is stated in `<locale>`?

I'm sorry, we do set the locale. Our `config.xml` includes:
```
<locale>de_DE</locale>
```
In fact, the `config.xml` is identical in both cases (42.3 vs. 15.0).

> Apart from the different behavior is there anything that is specially causing you an issue? Have you tested the output of `localectl` command?

The problem manifests itself such that special/accented characters in languages other than German (e.g. a French é) are displayed as '?'.

> Currently locale, keyboard and other related configurations are handled by systemd, so I would expect that the `localectl` returns you the correct value and that system behaves accordingly to systemd.

The output of `localectl` appears to be OK
```
builton150:~ # localectl 
   System Locale: LANG=de_DE.UTF-8
       VC Keymap: n/a
      X11 Layout: n/a
```
**however**, in order to fix (or workaround) the problem for us, we had to force `RC_LANG="de_DE.UTF-8"` in `/etc/sysconfig/language` (that was also the case when built on Leap 42.3). W/o the fix `RC_LANG` was empty, when built on Leap 15.0.

So perhaps a more accurate subject would have been 'Locale _partially_ unset if building ...'

**Comment by tilldoerges** [comment]
Oops. Closing was accidental.

**Comment by davidcassany** [comment]
@tilldoerges Ok the behaviour change is happening because of this 02dc8e32e1b619fc521ef9c0886d26079bcd3741 commit since version v9.12.0. This was introduced in the scope of #577, in which we were required to start using systemd tools to set locales starting from SLE15 (aka Leap 15) so in the code we assumed to start using `systemd-firstboot` tool having the old behavior as a fallback in case there was no `systemd-firstboot` tool with the appropriate capabilities.

The problem is how do we stablish  a criteria here https://github.com/SUSE/kiwi/blob/853efb4aecb9b35670b9165c709aa28696fb2e9b/kiwi/system/setup.py#L318-L333

I don't know how we could make the distinction of using one or the other strategy without having to start guessing the distro we are building.

In my opinion there isn't much we can do here, I don't see how can we find a safe and simple strategy. I'll give it a further thought though.

**Comment by tilldoerges** [comment]
How about changing the `elif` into an `if` in
```
elif os.path.exists(self.root_dir + '/etc/sysconfig/language'):
```
?

Apparently there can be multiple places to set the locale, so I'd say there's nothing wrong with setting all of them (or as many as possible).

**Comment by davidcassany** [comment]
@tilldoerges yes, this is the option I wanted to think about. I have been rereading the feature request we got back then. Unfortunately I can't remember all the discussions we had back then regarding this topic but I could not see anything against having this values set in both sites.

I am not a fan of setting those values in multiple places, probably there is no other option. However another fact is that support for Leap 42.3 is ending on 30 June 2019, so just two months are left. I am not fully convinced we should include code to workaround this regression that, theoretically, is about to disappear in two months time. 

Overall I would be fine by this change of setting locale in both sites, probably even an explanatory comment could be added in code, so next time we struggle with it we can decide and evaluate again weather to completely remove support for `/etc/sysconfig/language` or not.

@schaefi what do you think?

**Comment by tilldoerges** [comment]
> Overall I would be fine by this change of setting locale in both sites, probably even an explanatory comment could be added in code, so next time we struggle with it we can decide and evaluate again weather to completely remove support for `/etc/sysconfig/language` or not.

Does this mean `/etc/sysconfig/language` is not used on Leap 15.x anymore? Because, I briefly checked and found the file on several openSUSE Leap 15.0 boxes (but admittedly all of them were installed with `zypper dup` from previous openSUSE versions).

**Comment by davidcassany** [comment]
AFAIK systemd was updated to, at install or upgrade time, import values from `/etc/sysconfig/language` to the systemd managed files.

@tilldoerges I am wondering if your use case could be tested for Leap 15 image builds. Do you have a configuration to share and validate the issue is only present for Leap 42.3 images being build from Leap 15 hosts (or having kiwi >= 9.12.0)? I mean something like set locale to `<locale>de_DE</locale>`, add certain package and then run some command on the booted system to see locales inconsistency. Do you have some relevant for your use case that I could experiment with?

**Comment by tilldoerges** [comment]
Will try, but I'm not sure I can come up with a simple reproducer soonish.

**Comment by schaefi** [comment]
I'm late sorry... too many other things going on

Regarding the locale setup we use ```systemd-firstboot --locale``` and expected it would create all relevant entries. However it does create and manage /etc/locale.conf and /etc/vconsole.conf. Those files should list correct entries

The setup of the LANG / LC_* env variables which is responsible for the issue you see is usually done at boot time when systemd sets the locale. I was under the impression that /etc/sysconfig/language is of no relevant use anymore. That seems to be (partly) wrong

In my test build LC_CTYPE was set to the configured locale. LANG and all other LC_* vars were set to POSIX.

I would say it would be good if we run the code path that sets LANG in addition to the systemd-firstboot invocation.

**Comment by davidcassany** [comment]
> Will try, but I'm not sure I can come up with a simple reproducer soonish.

Ok, don't spent time in it. I just asked in case you had something simple at hand to verify with the same application as you do.

**Comment by davidcassany** [comment]
> I would say it would be good if we run the code path that sets LANG in addition to the systemd-firstboot invocation.

Agree gonna provide a patch.

With the current code I also verified that Leap 42.3 builds `<lcoale>` is not properly set, mostly because the systemd version present in Leap 42.3 already includes the locales options even not being used. We had a fallback code to use old LANG configs in environments where `systemd-firstboot` is not having the locale options available, but this is not the case for 42.3.

I also noted that Leap 15 locale settings only apply for non-root users, this relevant, as most of our tests only include the root user. I noted that root gets POSIX locale meanwhile other users get whatever is defined in `/etc/locale.conf`. This is also happening in my host (Leap 15) that was installed from an official installation media, thus I consider it to be the default openSUSE behavior.
